plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/IcebergSplitSource.java

```diff
@@ -178,74 +173,177 @@ public CompletableFuture<ConnectorSplitBatch> getNextBatch(ConnectorPartitionHan
+
+        Set<IcebergColumnHandle> identityPartitionColumns = partitionKeys.keySet().stream()
+                .map(fieldId -> getColumnHandle(fileSchema.findField(fieldId), typeManager))
+                .collect(toImmutableSet());
+
+        Supplier<Map<ColumnHandle, NullableValue>> partitionValues = memoize(() -> {
+            Map<ColumnHandle, NullableValue> bindings = new HashMap<>();
+            for (IcebergColumnHandle partitionColumn : identityPartitionColumns) {
+                Object partitionValue = deserializePartitionValue(
+                        partitionColumn.getType(),
+                        partitionKeys.get(partitionColumn.getId()).orElse(null),
+                        partitionColumn.getName());
+                NullableValue bindingValue = new NullableValue(partitionColumn.getType(), partitionValue);
+                bindings.put(partitionColumn, bindingValue);
             }
-            if (!partitionMatchesConstraint(identityPartitionColumns, partitionValues, constraint)) {
-                continue;
+            return bindings;
+        });
+
+        TupleDomain<IcebergColumnHandle> dynamicFilterPredicate = dynamicFilter.getCurrentPredicate()
+                .transformKeys(IcebergColumnHandle.class::cast);
+        if (!dynamicFilterPredicate.isAll() && !dynamicFilterPredicate.equals(pushedDownDynamicFilterPredicate)) {
+            if (!partitionMatchesPredicate(
+                    identityPartitionColumns,
+                    partitionValues,
+                    dynamicFilterPredicate)) {
+                return Optional.empty();
             }
-            if (recordScannedFiles) {
-                scannedFiles.add(scanTask.file());
+            if (!fileMatchesPredicate(
+                    fieldIdToType,
+                    dynamicFilterPredicate,
+                    scanTask.file().lowerBounds(),
+                    scanTask.file().upperBounds(),
+                    scanTask.file().nullValueCounts())) {
+                return Optional.empty();
             }
-            splits.add(icebergSplit);
         }
-        return completedFuture(new ConnectorSplitBatch(splits.build(), isFinished()));
+        if (!partitionMatchesConstraint(identityPartitionColumns, partitionValues, constraint)) {
+            return Optional.empty();
+        }
+        if (recordScannedFiles) {
+            scannedFiles.add(scanTask.file());
+        }
+        return Optional.of(toIcebergSplit(scanTask));
     }
 
     private void finish()
     {
         close();
-        this.combinedScanIterable = CloseableIterable.empty();
-        this.fileScanIterator = Collections.emptyIterator();
+        this.splitIterator = CloseableIterator.empty();
     }
 
     @Override
     public boolean isFinished()
     {
-        return fileScanIterator != null && !fileScanIterator.hasNext();
+        return splitIterator != null && !splitIterator.hasNext();
```

Comment:
finish has `this.splitIterator = CloseableIterator.empty();`

let's decide whether `splitIterator` is nullable and simplify finish or isfinished